lib/util/stream_utils.js

```diff
@@ -30,55 +30,117 @@ shaka.util.StreamUtils = class {
+    const variants = StreamUtils.filterVariantsByAudioChannelCount(
         manifest.variants, preferredAudioChannelCount);
 
-    /**
-     * @param {!shaka.extern.Variant} variant
-     * @return {string}
-     */
-    function variantCodecs(variant) {
-      // Only consider the base of the codec string.  For example, these should
-      // both be considered the same codec: avc1.42c01e, avc1.4d401f
-      let baseVideoCodec = '';
-      if (variant.video) {
-        baseVideoCodec = MimeUtils.getCodecBase(variant.video.codecs);
-      }
+    // Now organize variants into buckets by codecs.
+    /** @type {!shaka.util.MultiMap.<shaka.extern.Variant>} */
+    let variantsByCodecs = StreamUtils.getVariantsByCodecs(variants);
+    variantsByCodecs = StreamUtils.filterVariantsByHeight(variantsByCodecs);
+
+    const bestCodecs = StreamUtils.findBestCodecs(variantsByCodecs);
 
-      let baseAudioCodec = '';
-      if (variant.audio) {
-        baseAudioCodec = MimeUtils.getCodecBase(variant.audio.codecs);
+    // Filter out any variants that don't match, forcing AbrManager to choose
+    // from the most efficient variants possible.
+    manifest.variants = manifest.variants.filter((variant) => {
+      const codecs = StreamUtils.getGroupVariantCodecs(variant);
+      if (codecs == bestCodecs) {
+        return true;
       }
 
-      return baseVideoCodec + '-' + baseAudioCodec;
-    }
+      shaka.log.debug('Dropping Variant (better codec available)', variant);
+      return false;
+    });
+  }
 
-    // Now organize variants into buckets by codecs.
-    /** @type {!shaka.util.MultiMap.<shaka.extern.Variant>} */
+  /**
+  * Get variants by codecs.
+  *
+  * @param {!Array<shaka.extern.Variant>} variants
+  * @return {!shaka.util.MultiMap.<shaka.extern.Variant>}
+  */
+  static getVariantsByCodecs(variants) {
     const variantsByCodecs = new shaka.util.MultiMap();
     for (const variant of variants) {
-      const group = variantCodecs(variant);
+      const group = shaka.util.StreamUtils.getGroupVariantCodecs(variant);
       variantsByCodecs.push(group, variant);
     }
 
-    // Compute the average bandwidth for each group of variants.
-    // Choose the lowest-bandwidth codecs.
-    let bestCodecs = null;
+    return variantsByCodecs;
+  }
+
+  /**
+  * Filters variants with missing heights by all codecs.
+  *
+  * Average bandwidth of H.265 codecs (vp9) is higher than H.264 codecs (avc1).
+  * H.265 has 4K support and more variants.
+  *
+  * @param {!shaka.util.MultiMap.<shaka.extern.Variant>} variantsByCodecs
+  * @return {!shaka.util.MultiMap.<shaka.extern.Variant>}
+  */
+  static filterVariantsByHeight(variantsByCodecs) {
+    const heightByCodecs = new Map();
+    variantsByCodecs.forEach((codecs, variants) => {
+      for (const variant of variants) {
+        if (!variant.video) {
+          continue;
+        }
+
+        const key = variant.video.height;
```

Comment:
A few lines later, I forgot what "key" was.  How about we name this variable "height"?